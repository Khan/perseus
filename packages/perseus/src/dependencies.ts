import * as React from "react";

import type {PerseusDependencies, PerseusDependenciesV2} from "./types";

let _dependencies: PerseusDependencies | null | undefined = null;

export const setDependencies = (dependencies: PerseusDependencies) => {
    _dependencies = dependencies;
};

export const getDependencies = (): PerseusDependencies => {
    if (_dependencies) {
        return _dependencies;
    }

    throw new Error(
        [
            "Perseus has not been provided required dependencies.",
            "setDependencies(dependencies) must be called first.",
            "Make sure Perseus is being imported from javascript/perseus/perseus.js.",
        ].join("\n"),
    );
};

export type DependencyProps = Partial<
    React.ContextType<typeof DependenciesContext>
>;

export const DependenciesContext = React.createContext<PerseusDependenciesV2>({
    analytics: {onAnalyticsEvent: async () => {}},
});

/**
 * useDependencies provides easy access to the current Perseus dependencies.
 */
export const useDependencies = (): PerseusDependenciesV2 => {
    const deps = React.useContext(DependenciesContext);

    return deps;
};
