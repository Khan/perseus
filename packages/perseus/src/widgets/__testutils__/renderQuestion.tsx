import {RenderStateRoot} from "@khanacademy/wonder-blocks-core";
import {render} from "@testing-library/react";
import * as React from "react";

import {
    testDependenciesV2,
    testDependencies,
    // eslint-disable-next-line import/no-relative-packages
} from "../../../../../testing/test-dependencies";
import {
    DependenciesContext,
    useDependencies,
    setDependencies,
} from "../../dependencies";
import * as Perseus from "../../index";
import {mockStrings} from "../../strings";
import {registerAllWidgetsForTesting} from "../../util/register-all-widgets-for-testing";

import type {APIOptions} from "../../types";
import type {PerseusRenderer} from "@khanacademy/perseus-core";
import type {PropsFor} from "@khanacademy/wonder-blocks-core";

type RenderResult = ReturnType<typeof render>;

type ExtraProps = Omit<PropsFor<typeof Perseus.Renderer>, "strings">;

export const renderQuestion = (
    question: PerseusRenderer,
    apiOptions: APIOptions = Object.freeze({}),
    extraProps?: ExtraProps,
): {
    container: HTMLElement;
    renderer: Perseus.Renderer;
    rerender: (question: PerseusRenderer, extraProps?: ExtraProps) => void;
    unmount: RenderResult["unmount"];
} => {
    setDependencies(testDependencies);
    registerAllWidgetsForTesting();

    let renderer: Perseus.Renderer | null = null;
    const {container, rerender, unmount} = render(
        <RenderStateRoot>
            <DependenciesContext.Provider value={testDependenciesV2}>
                <Renderer
                    ref={(node) => (renderer = node)}
                    question={question as any}
                    apiOptions={apiOptions}
                    extraProps={{
                        ...extraProps,
                        strings: mockStrings,
                    }}
                />
            </DependenciesContext.Provider>
        </RenderStateRoot>,
    );
    // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
    if (!renderer) {
        throw new Error(`Failed to render!`);
    }
    const renderAgain = (
        question: PerseusRenderer,
        extraProps?: ExtraProps,
    ) => {
        rerender(
            <RenderStateRoot>
                <DependenciesContext.Provider value={testDependenciesV2}>
                    <Renderer
                        ref={(node) => (renderer = node)}
                        question={question}
                        apiOptions={apiOptions}
                        extraProps={{
                            ...extraProps,
                            strings: mockStrings,
                        }}
                    />
                </DependenciesContext.Provider>
            </RenderStateRoot>,
        );
        if (!renderer) {
            throw new Error(`Failed to rerender!`);
        }
    };

    return {container, renderer, rerender: renderAgain, unmount};
};

const Renderer = React.forwardRef<
    Perseus.Renderer,
    {
        question: PerseusRenderer;
        apiOptions: APIOptions;
        extraProps?: PropsFor<typeof Perseus.Renderer>;
    }
>((props, ref) => {
    const dependencies = useDependencies();
    return (
        <Perseus.Renderer
            ref={ref}
            content={props.question.content}
            images={props.question.images}
            widgets={props.question.widgets}
            problemNum={0}
            apiOptions={props.apiOptions}
            strings={mockStrings}
            {...props.extraProps}
            {...dependencies}
        />
    );
});
