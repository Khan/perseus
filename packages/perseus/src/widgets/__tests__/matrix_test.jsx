// @flow

import {screen} from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom";

import * as Dependencies from "../../../perseus-all-package/dependencies.js";
import {renderQuestion} from "../../../perseus-all-package/widgets/__tests__/renderQuestion.jsx";
import {testDependencies} from "../../../perseus-testing/test-dependencies.js";
import {question1} from "../__testdata__/matrix_testdata.js";

import type {APIOptions} from "../../../perseus-all-package/types.js";

describe("matrix widget", () => {
    beforeEach(() => {
        jest.spyOn(Dependencies, "getDependencies").mockReturnValue(
            testDependencies,
        );
    });

    it("should snapshot", () => {
        // Arrange
        const apiOptions: APIOptions = {
            isMobile: false,
        };

        // Act
        const {container} = renderQuestion(question1, apiOptions);

        // Assert
        expect(container).toMatchSnapshot("first render");
    });

    it("should snapshot on mobile", () => {
        // Arrange
        const apiOptions: APIOptions = {
            isMobile: true,
        };

        // Act
        const {container} = renderQuestion(question1, apiOptions);

        // Assert
        expect(container).toMatchSnapshot("first mobile render");
    });

    it("can be answered correctly", () => {
        // Arrange
        const apiOptions: APIOptions = {
            isMobile: false,
        };
        const {renderer} = renderQuestion(question1, apiOptions);

        // Act
        const correctAnswers = [5, -2, 1, 1, 1, 1, 7, -3, 3, 0, 0, -2];
        screen.getAllByRole("textbox").forEach((textbox, index) => {
            userEvent.paste(textbox, correctAnswers[index].toString());
        });

        renderer.guessAndScore();

        // assert
        expect(renderer).toHaveBeenAnsweredCorrectly();
    });

    it("can be answered incorrectly", () => {
        // Arrange
        const apiOptions: APIOptions = {
            isMobile: false,
        };
        const {renderer} = renderQuestion(question1, apiOptions);

        // Act
        screen.getAllByRole("textbox").forEach((textbox, index) => {
            userEvent.paste(textbox, "1");
        });

        renderer.guessAndScore();

        // Assert
        expect(renderer).toHaveBeenAnsweredIncorrectly();
    });
});
