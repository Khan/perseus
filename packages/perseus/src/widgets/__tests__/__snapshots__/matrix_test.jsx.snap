// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matrix widget should snapshot on mobile: first mobile render 1`] = `
<div>
  <span
    class="mock-KatexProvider"
  >
    <div
      class="perseus-renderer perseus-renderer-responsive"
    >
      <div
        class="paragraph"
        data-perseus-paragraph-index="0"
      >
        <div
          class="paragraph"
        >
          <strong>
            Perform the row operation, 
            <span
              style="white-space: nowrap;"
            >
              <span />
              <span
                class="mock-TeX"
              >
                R_3 \\leftrightarrow R_2
              </span>
              <span />
            </span>
            , on the following matrix.
          </strong>
        </div>
      </div>
      <div
        class="paragraph perseus-paragraph-centered"
        data-perseus-paragraph-index="1"
      >
        <div
          class="perseus-block-math"
          style="margin-left: -16px; margin-right: -16px;"
        >
          <div
            class="perseus-block-math-inner"
            style="overflow-x: auto; overflow-y: hidden; margin-top: -10px; margin-bottom: -10px; transform: translate3d(0,0,0); padding: 10px 16px 10px 16px;"
          >
            <span
              style="display: block; width: 100%; transform: scale(1, 1)  translate(0, 8px); transform-origin: 0 0; opacity: 0;"
            >
              <span
                class="mock-TeX"
              >
                \\left[\\begin{array} {ccc}
5 & -2 & 1 & 1 \\\\
3 & 0 & 0 & -2 \\\\
1 & 1 & 7 & -3 \\end{array} \\right] 
              </span>
            </span>
          </div>
        </div>
      </div>
      <div
        class="paragraph"
        data-perseus-paragraph-index="2"
      >
        <div
          class="paragraph"
        >
          <div
            class="perseus-widget-container widget-nohighlight widget-block"
          >
            <div
              class="perseus-matrix"
            >
              
              <div
                class="matrix-input"
              >
                <div
                  class="matrix-bracket bracket-left"
                  style="height: 36px;"
                />
                <div
                  class="matrix-bracket bracket-right"
                  style="height: 36px; left: 46px;"
                />
                <div
                  class="matrix-row"
                >
                  <span
                    class="matrix-input-field"
                  >
                    <input
                      class="inside"
                      style="height: 30px; width: 40px; margin: 3px;"
                      type="text"
                      value=""
                    />
                  </span>
                  <span
                    class="matrix-input-field"
                  >
                    <input
                      class="outside"
                      style="height: 30px; width: 40px; margin: 3px;"
                      type="text"
                      value=""
                    />
                  </span>
                  <span
                    class="matrix-input-field"
                  >
                    <input
                      class="outside"
                      style="height: 30px; width: 40px; margin: 3px;"
                      type="text"
                      value=""
                    />
                  </span>
                  <span
                    class="matrix-input-field"
                  >
                    <input
                      class="outside"
                      style="height: 30px; width: 40px; margin: 3px;"
                      type="text"
                      value=""
                    />
                  </span>
                </div>
                <div
                  class="matrix-row"
                >
                  <span
                    class="matrix-input-field"
                  >
                    <input
                      class="outside"
                      style="height: 30px; width: 40px; margin: 3px;"
                      type="text"
                      value=""
                    />
                  </span>
                  <span
                    class="matrix-input-field"
                  >
                    <input
                      class="outside"
                      style="height: 30px; width: 40px; margin: 3px;"
                      type="text"
                      value=""
                    />
                  </span>
                  <span
                    class="matrix-input-field"
                  >
                    <input
                      class="outside"
                      style="height: 30px; width: 40px; margin: 3px;"
                      type="text"
                      value=""
                    />
                  </span>
                  <span
                    class="matrix-input-field"
                  >
                    <input
                      class="outside"
                      style="height: 30px; width: 40px; margin: 3px;"
                      type="text"
                      value=""
                    />
                  </span>
                </div>
                <div
                  class="matrix-row"
                >
                  <span
                    class="matrix-input-field"
                  >
                    <input
                      class="outside"
                      style="height: 30px; width: 40px; margin: 3px;"
                      type="text"
                      value=""
                    />
                  </span>
                  <span
                    class="matrix-input-field"
                  >
                    <input
                      class="outside"
                      style="height: 30px; width: 40px; margin: 3px;"
                      type="text"
                      value=""
                    />
                  </span>
                  <span
                    class="matrix-input-field"
                  >
                    <input
                      class="outside"
                      style="height: 30px; width: 40px; margin: 3px;"
                      type="text"
                      value=""
                    />
                  </span>
                  <span
                    class="matrix-input-field"
                  >
                    <input
                      class="outside"
                      style="height: 30px; width: 40px; margin: 3px;"
                      type="text"
                      value=""
                    />
                  </span>
                </div>
              </div>
              
            </div>
          </div>
        </div>
      </div>
    </div>
  </span>
</div>
`;

exports[`matrix widget should snapshot: first render 1`] = `
<div>
  <span
    class="mock-KatexProvider"
  >
    <div
      class="perseus-renderer perseus-renderer-responsive"
    >
      <div
        class="paragraph"
        data-perseus-paragraph-index="0"
      >
        <div
          class="paragraph"
        >
          <strong>
            Perform the row operation, 
            <span
              style="white-space: nowrap;"
            >
              <span />
              <span
                class="mock-TeX"
              >
                R_3 \\leftrightarrow R_2
              </span>
              <span />
            </span>
            , on the following matrix.
          </strong>
        </div>
      </div>
      <div
        class="paragraph perseus-paragraph-centered"
        data-perseus-paragraph-index="1"
      >
        <div
          class="perseus-block-math"
        >
          <div
            class="perseus-block-math-inner"
            style="overflow-x: auto; overflow-y: hidden; padding-top: 10px; padding-bottom: 10px; margin-top: -10px; margin-bottom: -10px;"
          >
            <span
              class="mock-TeX"
            >
              \\left[\\begin{array} {ccc}
5 & -2 & 1 & 1 \\\\
3 & 0 & 0 & -2 \\\\
1 & 1 & 7 & -3 \\end{array} \\right] 
            </span>
          </div>
        </div>
      </div>
      <div
        class="paragraph"
        data-perseus-paragraph-index="2"
      >
        <div
          class="paragraph"
        >
          <div
            class="perseus-widget-container widget-nohighlight widget-block"
          >
            <div
              class="perseus-matrix"
            >
              
              <div
                class="matrix-input"
              >
                <div
                  class="matrix-bracket bracket-left"
                  style="height: 36px;"
                />
                <div
                  class="matrix-bracket bracket-right"
                  style="height: 36px; left: 46px;"
                />
                <div
                  class="matrix-row"
                >
                  <span
                    class="matrix-input-field"
                  >
                    <input
                      class="inside"
                      style="height: 30px; width: 40px; margin: 3px;"
                      type="text"
                      value=""
                    />
                  </span>
                  <span
                    class="matrix-input-field"
                  >
                    <input
                      class="outside"
                      style="height: 30px; width: 40px; margin: 3px;"
                      type="text"
                      value=""
                    />
                  </span>
                  <span
                    class="matrix-input-field"
                  >
                    <input
                      class="outside"
                      style="height: 30px; width: 40px; margin: 3px;"
                      type="text"
                      value=""
                    />
                  </span>
                  <span
                    class="matrix-input-field"
                  >
                    <input
                      class="outside"
                      style="height: 30px; width: 40px; margin: 3px;"
                      type="text"
                      value=""
                    />
                  </span>
                </div>
                <div
                  class="matrix-row"
                >
                  <span
                    class="matrix-input-field"
                  >
                    <input
                      class="outside"
                      style="height: 30px; width: 40px; margin: 3px;"
                      type="text"
                      value=""
                    />
                  </span>
                  <span
                    class="matrix-input-field"
                  >
                    <input
                      class="outside"
                      style="height: 30px; width: 40px; margin: 3px;"
                      type="text"
                      value=""
                    />
                  </span>
                  <span
                    class="matrix-input-field"
                  >
                    <input
                      class="outside"
                      style="height: 30px; width: 40px; margin: 3px;"
                      type="text"
                      value=""
                    />
                  </span>
                  <span
                    class="matrix-input-field"
                  >
                    <input
                      class="outside"
                      style="height: 30px; width: 40px; margin: 3px;"
                      type="text"
                      value=""
                    />
                  </span>
                </div>
                <div
                  class="matrix-row"
                >
                  <span
                    class="matrix-input-field"
                  >
                    <input
                      class="outside"
                      style="height: 30px; width: 40px; margin: 3px;"
                      type="text"
                      value=""
                    />
                  </span>
                  <span
                    class="matrix-input-field"
                  >
                    <input
                      class="outside"
                      style="height: 30px; width: 40px; margin: 3px;"
                      type="text"
                      value=""
                    />
                  </span>
                  <span
                    class="matrix-input-field"
                  >
                    <input
                      class="outside"
                      style="height: 30px; width: 40px; margin: 3px;"
                      type="text"
                      value=""
                    />
                  </span>
                  <span
                    class="matrix-input-field"
                  >
                    <input
                      class="outside"
                      style="height: 30px; width: 40px; margin: 3px;"
                      type="text"
                      value=""
                    />
                  </span>
                </div>
              </div>
              
            </div>
          </div>
        </div>
      </div>
    </div>
  </span>
</div>
`;
